nominal/core/_utils.py

```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from typing import Iterable, Protocol, TypeVar, runtime_checkable
+
+from .._utils import logger
+
+T = TypeVar("T")
+
+
+@runtime_checkable
+class HasRid(Protocol):
+    rid: str
+
+
+def rid_from_instance_or_string(value: HasRid | str) -> str:
+    if isinstance(value, str):
+        return value
+    elif isinstance(value, HasRid):
+        return value.rid
+    raise TypeError("{value!r} is not a string nor an instance with a 'rid' attribute")
```

Comment:
Missing f for format string?

Reply:
good catch! should get a linter running at some point soon.